akka-http-core/src/main/scala/akka/http/impl/engine/server/ServerTerminator.scala

```diff
@@ -0,0 +1,294 @@
+
+  val fromUser: Inlet[HttpResponse] = Inlet("userIn")
+  val toNet: Outlet[HttpResponse] = Outlet("netOut")
+  override def shape = BidiShape(fromUser, toNet, fromNet, toUser)
+
+  final val TerminationDeadlineTimerKey = "TerminationDeadlineTimerKey"
+
+  final class ConnectionTerminator(triggerTermination: Promise[FiniteDuration ⇒ Future[HttpTerminated]]) extends ServerTerminator {
+    override def terminate(deadline: FiniteDuration)(implicit ec: ExecutionContext): Future[HttpTerminated] =
+      triggerTermination.future.flatMap(callback ⇒ callback(deadline))
+  }
+
+  override def createLogicAndMaterializedValue(inheritedAttributes: Attributes): (GraphStageLogic, ServerTerminator) = {
+    val triggerTermination = Promise[FiniteDuration ⇒ Future[HttpTerminated]]() // result here means termination of this connection has completed
+
+    // responsible for terminating this connection
+    val selfTerminator = new ConnectionTerminator(triggerTermination)
+
+    val logic = new TimerGraphStageLogic(shape) with StageLogging {
+
+      // this promise will be completed once our termination is complete;
+      // e.g. we replied with "go away" to pending requests, and no new ones were incoming etc
+      val terminationOfConnectionDone = Promise[HttpTerminated]() // FIXME this is a wrong name... it is only the connection that terminated here
+
+      // true, if a request was delivered to the user-handler, and no response was sent yet
+      // in that case, if the termination timeout triggers, we will need to render a synthetic response to the client.
+      var pendingResponse: Boolean = false
+
+      override def preStart(): Unit = {
+        val terminateSignal = getAsyncCallback[FiniteDuration] { deadline ⇒
+          log.debug("[terminator] Initializing termination of server, deadline: {}", PrettyDuration.format(deadline))
+          installTerminationHandlers(deadline.fromNow)
+
+          scheduleOnce(TerminationDeadlineTimerKey, deadline)
+        }
+        triggerTermination.success { d ⇒
+          terminateSignal.invoke(d)
+          terminationOfConnectionDone.future
+        }
+      }
+
+      setHandler(fromUser, new InHandler {
+        override def onPush(): Unit = {
+          val response = grab(fromUser)
+          pendingResponse = false
+          push(toNet, response)
+        }
+      })
+      setHandler(toUser, new OutHandler {
+        override def onPull(): Unit =
+          pull(fromNet)
+      })
+      setHandler(fromNet, new InHandler {
+        override def onPush(): Unit = {
+          val request = grab(fromNet)
+          pendingResponse = true
+          push(toUser, request)
+        }
+      })
+      setHandler(toNet, new OutHandler {
+        override def onPull(): Unit = {
+          pull(fromUser)
+        }
+      })
+
+      // TODO what other semantics should we change in this phase...
+      def installTerminationHandlers(deadline: Deadline): Unit = {
```

Comment:
some semantics debate could happen in those handlers... those seem correct AFAICS but maybe more tests needed